packages/collaboration-manager/src/Operation.ts

```diff
@@ -56,4 +56,130 @@ export class Operation {
     this.index = index;
     this.data = data;
   }
+
+  /**
+   * Makes an inverse operation
+   */
+  public inverse(): Operation {
+    const index = this.index;
+
+    switch (this.type) {
+      case OperationType.Insert:
+
+        const textRange = index.textRange;
+
+        if (textRange == undefined) {
+          throw new Error('Unsupported index');
+        }
+
+        const [ textRangeStart ] = textRange;
+
+        const newIndex = new IndexBuilder()
+          .from(index)
+          .addTextRange([textRangeStart, textRangeStart + this.data.newValue.length])
+          .build();
+
+        return new Operation(OperationType.Delete, newIndex, {
+          prevValue: this.data.newValue,
+          newValue: this.data.prevValue,
+        });
+      case OperationType.Delete:
+        return new Operation(OperationType.Insert, index, {
+          prevValue: this.data.newValue,
+          newValue: this.data.prevValue,
+        });
+      case OperationType.Modify:
+        return new Operation(OperationType.Modify, index, {
+          prevValue: this.data.newValue,
+          newValue: this.data.prevValue,
+        });
+    }
+  }
+
+  /**
+   * Transforms the operation against another operation
+   *
+   * @param againstOp - operation to transform against
+   */
+  public transform(againstOp: Operation): Operation {
+    if (!this.index.isTextIndex || !againstOp.index.isTextIndex) {
+      throw new Error('Unsupported index');
+    }
+
+    if (this.type === OperationType.Modify || againstOp.type === OperationType.Modify) {
+      throw new Error('Unsupported operation type');
+    }
+
+    /**
+     * Do not transform operations if they are on different blocks or documents
+     */
+    if (this.index.documentId !== againstOp.index.documentId || this.index.blockIndex !== againstOp.index.blockIndex) {
+      return this;
+    }
+
+    const [ receivedStartIndex ] = this.index.textRange!;
+    const [ localStartIndex ] = againstOp.index.textRange!;
+
+    switch (true) {
+      case this.type === OperationType.Insert && againstOp.type === OperationType.Insert:
```

Comment:
The type of the current operation doesn't matter here, index will be changed the same way for both insert and delte

Reply:
There are only three cases here:
1. If received operation is delete operatio **before** (index-wise) the current, we need to update the current one index shifting it left
2. If received operation is insert operation **before** (index-wise) the current, we need to shift index right
3. If operation of any type happened **after** (index-wise) the current, no need to do anything